### PR TITLE
Further Elastic.Ingest Improvements

### PR DESCRIPTION
**Pull request by Mpdreamz**
This builds upon #109 where `Elastic.Ingest` moves to the product neutral dependency `Elastic.Transport`. 

This PR introduces the following:

* `Elastic.Ingest` is now the base library for anything that wants to ship buffered data partioned by buffer size and max buffer age.
* `Elastic.Ingest.Transport` is now the base for anything that wants to push data using `Elastic.Transport`
	  * `Elastic.Ingest.Elasticsearch` allows any object to pushed to any index.
	  	 This PR introduces 2 Elasticsearch channels `IndexChannel` and a `DataStreamChannel` to provide the best experience for both without trying to abstract a single channel to work for both.
	  * `Elastic.Ingest.Apm` only allows apm events to be pushed to apm server.
* `Elastic.Ingest.OpenTelemetry` is a toy library that allows you to write `Activity`'s to an OpenTelemetry receiver using `TraceChannel` 

This PR also splits the internal channel into an inbound and outbound channel. The inbound channel's sole purpose it to batch based on size/lifetime. It then sends pages of data to the outbound channel where data is sequentially received and send out concurrently. 

Example usage of `ApmChannel` pushing transactions:

```csharp

var channelOptions = new ApmChannelOptions
{
	BufferOptions = options,
	ShipTo = new ShipTo(transport)
};

var channel = new ApmChannel(channelOptions);

string Id() => RandomGenerator.GenerateRandomBytesAsString(8);
var random = new Random();

for (var i = 0; i < numberOfEvents; i++)
{
	channel.TryWrite(new Transaction("http", Id(), Id(), new SpanCount(), random.NextDouble() * random.Next(100, 1000) , Epoch.UtcNow) { Name = "x" });
}
```

image

Opening early for discussing the approach, plenty of refactoring and actual implementations need to follow.

### CODE REVIEW

**Comment by Mpdreamz** [comment]
This is currently blocked due to a weird nuget restore error on the windows build machine:

```
[2022-08-11T15:09:58.038Z] C:\Users\jenkins\workspace\ent-dotnet_ecs-dotnet-mbp_PR-110\ecs-dotnet\build\scripts\scripts.fsproj : error NU1101: Unable to find package MinVer. No packages exist with this id in source(s): C:\Program Files\dotnet\sdk\6.0.302\FSharp\library-packs

[2022-08-11T15:09:58.039Z] C:\Users\jenkins\workspace\ent-dotnet_ecs-dotnet-mbp_PR-110\ecs-dotnet\build\scripts\scripts.fsproj : error NU1101: Unable to find package Microsoft.NETFramework.ReferenceAssemblies.net461. No packages exist with this id in source(s): C:\Program Files\dotnet\sdk\6.0.302\FSharp\library-packs

[2022-08-11T15:09:58.039Z] C:\Users\jenkins\workspace\ent-dotnet_ecs-dotnet-mbp_PR-110\ecs-dotnet\build\scripts\scripts.fsproj : error NU1101: Unable to find package Argu. No packages exist with this id in source(s): C:\Program Files\dotnet\sdk\6.0.302\FSharp\library-packs

[2022-08-11T15:09:58.039Z] C:\Users\jenkins\workspace\ent-dotnet_ecs-dotnet-mbp_PR-110\ecs-dotnet\build\scripts\scripts.fsproj : error NU1101: Unable to find package Bullseye. No packages exist with this id in source(s): C:\Program Files\dotnet\sdk\6.0.302\FSharp\library-packs

[2022-08-11T15:09:58.039Z] C:\Users\jenkins\workspace\ent-dotnet_ecs-dotnet-mbp_PR-110\ecs-dotnet\build\scripts\scripts.fsproj : error NU1101: Unable to find package Proc. No packages exist with this id in source(s): C:\Program Files\dotnet\sdk\6.0.302\FSharp\library-packs

[2022-08-11T15:09:58.040Z] C:\Users\jenkins\workspace\ent-dotnet_ecs-dotnet-mbp_PR-110\ecs-dotnet\build\scripts\scripts.fsproj : error NU1101: Unable to find package Fake.Tools.Git. No packages exist with this id in source(s): C:\Program Files\dotnet\sdk\6.0.302\FSharp\library-packs
```

Something I am unable to reproduce locally.

**Comment by Mpdreamz** [comment]
@gregkalapos this is now ready for review

File: src/Elastic.Ingest.Apm/ApmChannel.cs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+			// "system":{"architecture":"amd64","detected_hostname":"8ec7ceb99074","configured_hostname":"host1","platform":"Linux","container":{"id":"8ec7ceb990749e79b37f6dc6cd3628633618d6ce412553a552a0fa6b69419ad4"},
+			// "kubernetes":{"namespace":"default","pod":{"uid":"b17f231da0ad128dc6c6c0b2e82f6f303d3893e3","name":"instrumented-java-service"},"node":{"name":"node-name"}}},
+			// "service":{"name":"1234_service-12a3","version":"4.3.0","node":{"configured_name":"8ec7ceb990749e79b37f6dc6cd3628633618d6ce412553a552a0fa6b69419ad4"},"environment":"production","language":{"name":"Java","version":"10.0.2"},
+			// "agent":{"version":"1.10.0","name":"java","ephemeral_id":"e71be9ac-93b0-44b9-a997-5638f6ccfc36"},"framework":{"name":"spring","version":"5.0.0"},"runtime":{"name":"Java","version":"10.0.2"}},"labels":{"group":"experimental","ab_testing":true,"segment":5}}}
+			// TODO cache
+			var p = Process.GetCurrentProcess();
+			var metadata = new { metadata = new { process = new { pid = p.Id, title = p.ProcessName }, service = new { name = p.ProcessName, version = "1.0.0", agent = new { name = "dotnet", version = "0.0.1"} } } };
```
Comment (gregkalapos) [inline src/Elastic.Ingest.Apm/ApmChannel.cs:71]:
```suggestion
			var metadata = new { metadata = new { process = new { pid = p.Id, title = p.ProcessName }, service = new { name  = System.Text.RegularExpressions.Regex.Replace(p.ProcessName, "[^a-zA-Z0-9 _-]", "_"), version = "1.0.0", agent = new { name = "dotnet", version = "0.0.1"} } } };
```

We have the same in the agent code - there is validation on the service name e.g. having a `.` in it is not accepted by the APM Server and that would make all the events rejected.

**Review by gregkalapos — APPROVED** [review]
Very nice! 🎉 

I added some comments, none of those are meant as a blocker or anything... just a few thoughts.

**Comment by Mpdreamz** [comment]
Feels amazing to merge a 2 year old PR :partying_face: :tada: . Thanks @gregkalapos for giving this massive PR a review!

**Pull request merged**